### PR TITLE
Updating metrics and checks

### PR DESCRIPTION
**Pull request by andrewma2**
## What changes are proposed in this pull request?

Add new metrics and update query checking to cover edge case. Also updated unit test code

See [this dashboard](https://grafana.cloud.databricks.com/d/jfb0fkenz/m3coordinator-cache-monitoring?orgId=1&from=now-3h&to=now&refresh=30s&var-datasource=dev-aws-us-west-1) for the updated metrics

## How is this tested?

Deployed to `dev-aws-us-west-1` and verified accuracy and new metrics recording

#### Code Review
For information about the code review process, e.g. how to find a reviewer or how to ping non-responsive reviewers, check the contents of [go/code-review](http://go/code-review) Confluence page.

#### Approvals
Other than the mandatory approvers enforced by the OWNER file framework (http://go/owners), this PR
requires at least one approval from another engineer.

#### [NEW] Shiproom

**Platform & Compute Fabric**:
Changes should be tracked by an approved "material change." Multiple PRs may be tracked by a single material change.

- [ ] Change modifies code owned or released by a Platform or Compute Fabric team
  - [ ] A "material change" covering this PR exists in http://go/engshiproom: `<CHANGE_ID>`

See http://go/platshiproomwiki for instructions and use http://go/lightcm-template to evaluate risk. Ask questions in #platform-shiproom.

**Runtime changes**:
- [ ] Change targets a runtime maintenance release (i.e., targets a maintenance `dbr-branch-x.x` branch or has a maintenance `dbr-branch-x.x` label)
  - [ ] Change is **NOT** a “material change”
  - [ ] Change **IS** a “material change” of low / medium risk
  - [ ] Change **IS** a “material change” of high risk needing Shiproom review

Please refer Runtime Shiproom Wiki: http://go/runtimeshiproomwiki

#### Security implications
_This section is intended for the reviewers of this PR_
Please, make sure you consider the content of "What are the responsibilities of code reviewers?" section of [go/pr-security-review](http://go/pr-security-review)

### CODE REVIEW

File: src/query/storage/cache/utils.go [hunk 1]
```diff
@@ -98,22 +98,57 @@ func samplesEqual(a, b []prompb.Sample, end int64, logger *zap.Logger) bool {
 	return true
 }
 
+// Create new result that doesn't include data past end
+func createEndFilteredTimeseries(timeseries []*prompb.TimeSeries, end int64) []*prompb.TimeSeries {
```
Comment (davidyuanfs) [inline src/query/storage/cache/utils.go:102]: looks like many copies in this method, would this cause high cpu or memory?

File: src/query/storage/cache/redis_cache.go [hunk 2]
```diff
@@ -98,10 +98,17 @@ type CacheMetrics struct {
...
+	cacheBytesRetrieved tally.Histogram
+	m3dbBytesRetrieved  tally.Histogram
 }
 
 func NewCacheMetrics(scope tally.Scope) CacheMetrics {
 	subScope := scope.SubScope("cache")
+	// Histogram buckets range from 10^3 to 10^12 (scaled by a factor of 10 each time)
+	// Bucketting chosen to have enough buckets and a wide enough range for
+	// histogram_quantile() to be reliable
+	buckets, _ := tally.ExponentialValueBuckets(1000, 10, 10)
```
Comment (davidyuanfs) [inline src/query/storage/cache/redis_cache.go:116]: does this include p99/p90 and p50 and p10?

File: src/query/storage/cache/utils.go [hunk 3]
```diff
@@ -98,22 +98,57 @@ func samplesEqual(a, b []prompb.Sample, end int64, logger *zap.Logger) bool {
...
+}
+
 // Check if lists of timeseries are equal (see samplesEqual for usage of {end} param)
 func TimeseriesEqual(a, b []*prompb.TimeSeries, end int64, logger *zap.Logger) bool {
-	if len(a) != len(b) {
-		logger.Info("Timeseries lengths unequal", zap.Int("actual length", len(a)), zap.Int("expected length", len(b)))
+	// Account for case where a new timeseries has just been recently added in the time difference
+	// In this case, filter out that data and recompare
+	a_filter := createEndFilteredTimeseries(a, end)
+	b_filter := createEndFilteredTimeseries(b, end)
```
Comment (davidyuanfs) [inline src/query/storage/cache/utils.go:141]: Actually when we skip the up to date minute to do the check, would be new metrics still concern?

File: src/query/storage/cache/redis_cache.go [hunk 4]
```diff
@@ -142,17 +153,19 @@ func (cm CacheMetrics) CacheMetricsBucketHit(results []*storage.PromResult) {
...
 func (cm CacheMetrics) CacheMetricsMiss(result storage.PromResult) {
-	// cm.missCounter.Inc(1)
+	cm.missCounter.Inc(1)
 	tot_samples := 0
 	for _, ts := range result.PromResult.Timeseries {
 		tot_samples += len(ts.Samples)
 	}
 	cm.missSamplesCounter.Inc(int64(tot_samples))
 	cm.missBytesCounter.Inc(int64(result.PromResult.Size()))
+	cm.m3dbBytesRetrieved.RecordValue(float64(result.PromResult.Size()))
```
Comment (davidyuanfs) [inline src/query/storage/cache/redis_cache.go:179]: I think this is not per simple byte, this is per request bytes right?

File: src/query/storage/cache/utils.go (hunk 1 above)
Comment (andrewma2) [reply src/query/storage/cache/utils.go:102]: Since the check rate is pretty low (0.5% of all requests), I don't think it should lead to high CPU usage or memory usage

File: src/query/storage/cache/redis_cache.go (hunk 2 above)
Comment (andrewma2) [reply src/query/storage/cache/redis_cache.go:116]: I can use the histogram_quantile() prometheus function to specify any quantile I think, see [dashboard](https://grafana.cloud.databricks.com/d/jfb0fkenz/m3coordinator-cache-monitoring?orgId=1&from=now-3h&to=now&refresh=30s&var-datasource=dev-aws-us-west-1) for P50, P90, P95, and P99 currently

File: src/query/storage/cache/utils.go (hunk 3 above)
Comment (andrewma2) [reply src/query/storage/cache/utils.go:141]: So the first check I make is that the number of metrics is the same so this is just a way to work around that check in case we add a new metric

File: src/query/storage/cache/redis_cache.go (hunk 4 above)
Comment (andrewma2) [reply src/query/storage/cache/redis_cache.go:179]: Yeah this metric is to record the number of bytes for each request. I already record the total number of bytes as a whole (missBytesCounter)

File: src/query/storage/cache/utils.go (hunk 1 above)
Comment (andrewma2) [reply src/query/storage/cache/utils.go:102]: Just checked with profiling, checking doesn't seem to contribute to much CPU usage

File: src/query/storage/cache/utils.go (hunk 3 above)
Comment (davidyuanfs) [reply src/query/storage/cache/utils.go:141]: But we already skip the first minute right? If we don't check the first minute, will we have new metrics?

Comment (andrewma2) [reply src/query/storage/cache/utils.go:141]: So the part where I disregard the first minute comes later in that code. But before that I added something to check that the number of metrics are equal since this made comparison easier. The problem is that if a new metric only exists in that first minute for the newer request, it won't pass this metric number check and so it will fail before I do the check that skips over the first minute. This is just a way to make sure that this case makes it past that check for equal metric counts

**Review by davidyuanfs — APPROVED** [review]

**Review by jnyi — COMMENTED** [review]
lgtm
